lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp

```diff
@@ -1694,6 +1694,90 @@ bool circt::firrtl::scatterCustomAnnotations(
       continue;
     }
 
+    if (clazz == "sifive.enterprise.grandcentral.ModuleReplacementAnnotation") {
+      auto id = newID();
+      NamedAttrList fields;
+      auto annotationsAttr =
+          tryGetAs<ArrayAttr>(dict, dict, "annotations", loc, clazz);
+      auto circuitAttr =
+          tryGetAs<StringAttr>(dict, dict, "circuit", loc, clazz);
+      auto circuitPackageAttr =
+          tryGetAs<StringAttr>(dict, dict, "circuitPackage", loc, clazz);
+      auto dontTouchesAttr =
+          tryGetAs<ArrayAttr>(dict, dict, "dontTouches", loc, clazz);
+      if (!annotationsAttr || !circuitAttr || !circuitPackageAttr || !dontTouchesAttr)
+        return false;
+      fields.append("class", classAttr);
+      fields.append("id", id);
+      fields.append("annotations", annotationsAttr);
+      fields.append("circuit", circuitAttr);
+      fields.append("circuitPackage", circuitPackageAttr);
+      newAnnotations["~"].push_back(DictionaryAttr::get(context, fields));
```

Comment:
I was surprised that this did _not_ include a `targets` field in the circuit-level annotation.  However, all the information in there is already handled based on the ordering of the target IDs.

This is fine, but will miss a situation where the _last_ target was deleted... e.g., if the targets are `[4, 5, 6]` and `6` gets deleted by a pass, then this would be lost. This can either be fixed by including a `numTargets` field here or by just doing a `targets = [4, 5, 6]` array here.

What do you think?

Reply:
I don't think module replacement annotation needs to keep track of deleted targets? in SFC we don't really care about which ones get deleted, the update method just silently removes targets from the list if they get deleted.

Reply:
If silently dropping deleted modules/instances is the correct behavior, then not recording the targets in the circuit-level annotation is the most efficient solution. 💯